### PR TITLE
Sign calculator

### PR DESCRIPTION
**Pull request by olim88**
adds ability to do calculations when inputting number into signs and a calculate command

# sign input
image
image

# command
image
image

I have created custom calculator utility for this using a shunting yard algorithm to convert calculation to RPN and then calculating the answer. it also supports (s, k, m, b) and created tests for this.

also tested in game bazaar and ah

### CODE REVIEW

**Comment by viciscat** [comment]
Another banger by olim. Do you think it would be doable to have an interface that a screen could implement to have it somewhere else? i.e. the popup to set the price/bid in fancy AH.

**Comment by olim88** [comment]
>  implement to have it somewhere else

This should not be to hard. If the renderSign method in SignCalculator to a location (add possible renamed) then it should work were ever this is called. is this the sort of thing you meant

**Comment by Fluboxer** [comment]
add "e" for "enchanted", equals 160 (2.5 stacks)

**Comment by viciscat** [comment]
> This should not be to hard. If the renderSign method in SignCalculator to a location (add possible renamed) then it should work were ever this is called. is this the sort of thing you meant

yea that's good enough

**Comment by kevinthegreat1** [comment]
Maybe only works if you have a `=` sign in front like `=4s * 5`. Also would be nice to have `purse` be the number of coins you have in your purse.

File: src/main/java/de/hysky/skyblocker/skyblock/calculators/SignCalculator.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    private static final MinecraftClient CLIENT = MinecraftClient.getInstance();
+
+    private static final DecimalFormat FORMATTER = new DecimalFormat("#,###.##");
+
+    private static String lastInput;
+    private static String input;
+    private static Double output;
+
+    public static void renderSign(DrawContext context, String[] messages) {
+        input = messages[0];
```
Comment (viciscat) [inline src/main/java/de/hysky/skyblocker/skyblock/calculators/SignCalculator.java:22]: Passing a string array only to use the first one is kinda pointless in my opinion. Just pass a normal string

**Review by viciscat — CHANGES_REQUESTED** [review]

**Comment by olim88** [comment]
> add "e" for "enchanted"
> have purse be the number of coins you have in your purse

good idea. i have added this and put a key in the tooltip so players know its there
> Maybe only works if you have a = sign in front

i feel like this is a nice idea and some will like it however i like to be reassured what the value of the sign is going to be even if i am just doing something like 1.02k so i have made it a toggle in the settings so both options are available

**Comment by viciscat** [comment]
last nitpick and I think this PR will be good, add "p" for "purse" maybe? to make it easier and faster to type

**Comment by Fluboxer** [comment]
> last nitpick and I think this PR will be good, add "p" for "purse" maybe? to make it easier and faster to type

I still don't get how this one would be useful tho

**Comment by viciscat** [comment]
> I still don't get how this one would be useful tho

make it faster to type or something, faster to type "p" than "purse"

File: src/main/resources/assets/skyblocker/lang/en_us.json [hunk 2]
```diff
@@ -537,6 +537,13 @@
   "skyblocker.config.uiAndVisuals.searchOverlay.maxSuggestions": "Maximum Suggestions",
   "skyblocker.config.uiAndVisuals.searchOverlay.maxSuggestions.@Tooltip": "The maximum number of suggested items to show.",
 
+  "skyblocker.config.uiAndVisuals.inputCalculator": "Input Calculator",
```
Comment (AzureAaron) [inline src/main/resources/assets/skyblocker/lang/en_us.json:540]: Ensure this is in alphabetical order with the rest of the translation strings

**Review by AzureAaron — DISMISSED** [review]

**Review by kevinthegreat1 — APPROVED** [review]
There's a tiny bug when you have `requiresEquals` set to true and type a valid equation without the equals sign. When you add the equals sign, it will show as invalid since the equation is not reevaluated when an equals sign is added, but I don't see a good way of fixing this.

**Pull request merged**